ADR: Rclone and Encryption adheres Cryptography Policy (#658)
elastisys/compliantkubernetes: docs/adr/0043-rclone-and-encryption-adhere-cryptography-policy.md

# Rclone and Encryption adheres Cryptography Policy

* Status: Accepted
* Deciders: Arch meeting
* Date: 2023-05-04

## Context and Problem Statement

> We want to enable rclone for environments in cleura public cloud including AMS's buckets to `sto1hs` in compliant cluera cloud.

Reviewer: Can we rephrase the Context and Problem Statement to make it more general to the Product and not to our Operations?

I would write something like:

```
Compliant Kubernetes can be configured to replicate the primary backup to a secondary backup on a second object storage. Under the hood, this is performed using `rclone`. Such replication improves disaster recovery and mitigates ransomware attacks.

However, this second object storage sometimes needs to be on a different cloud and/or in a different country.
In such cases, it is desirable to make the secondary backup completely opaque to the cloud provider, so that they do not count as a new data sub-processor.

Is `rclone`'s encryption sufficient for our purposes? Does it comply with the recommended [cryptography policy](../operator-manual/cryptography/)?
```

Also if decided, to run rclone to another region of Cluera public cloud, we wanted to consider using the the encryption feature in `rclone` to encrypt backups before they are sent to an off-site location outside of Sweden and it follows the standards set by our [Cryptography Policy](./../operator-manual/cryptography.md)

## Decision Drivers

* We want to maintain platform security and stability.

> * We want to find a solution which is scalable and minimizes MSE burden.

Reviewer: Let's consistently use our [Glossary](https://elastisys.io/compliantkubernetes/glossary/).

```suggestion
* We want to find a solution which is scalable and minimizes platform administrator burden.
```

* We want to best serve the Application Developers.
* We want to make the Platform Administrator life easier.
* We want to avoid cloud-provider dependent implementation sprawl.

## Considered Options

> 1. Allow-list the Public Cloud VM's ips.

Reviewer: This sounds like we're mixing an operational decision with an architectural decision. Could we separate the two?

Ideally, ADR should be written in a way that even the onprem team and outside parties can benefit from them.

# Rclone and Encryption adheres Cryptography Policy

* Status: Accepted
* Deciders: Arch Meeting
* Date: 2023-05-04

## Context and Problem Statement

Compliant Kubernetes can be configured to replicate the primary backup to a secondary backup on a second object storage. Under the hood, this is performed using `rclone`. Such replication improves disaster recovery and mitigates ransomware attacks.

> However, this second object storage sometimes needs to be on a different cloud and/or in a different country.

Reviewer 2: Replace `cloud` and `cloud provider` with `Infrastructure Provider` so we are consistent with our Glossary.

Author: Agreed, thanks @lucianvlad done the changes.

In such cases, it is desirable to make the secondary backup completely opaque to the cloud provider, so that they do not count as a new data sub-processor.

Is `rclone`'s encryption sufficient for our purposes? Does it comply with the recommended [cryptography policy](../operator-manual/cryptography.md)?

## Decision Drivers

* We want to maintain Platform security and stability.
* We want to find a solution which is scalable and minimizes Platform Administrator burden.
* We want to best serve the Application Developers.
* We want to make the Platform Administrator life easier.
* We want to avoid cloud-provider dependent implementation sprawl.

## Considered Options

> 1. Use `rclone` to replicate the data from our `primary` cloud provider to `secondary` public cloud provider on a second object storage.

Reviewer 2: Let's try to stay in line with our new terminology from the glossary and change `cloud provider` to `Infrastructure Provider` in all the lines it is used.

Author: Agreed, thanks @lucianvlad done the changes.

    - `Good`, because it supports our goals of data redundancy, privacy, compliance, and cross-cloud flexibility.

>     - `Bad`, because in some jurisdictions and regulatory frameworks, cloud providers might be considered data sub-processors, which could impose additional requirements and compliance obligations.

Reviewer 2: Same here..
Let's try to stay in line with our new terminology from the glossary and change `cloud provider` to `Infrastructure Provider` in all the lines it is used.

Author: Agreed, thanks @lucianvlad done the changes.

> 2. Use `rclone` to replicate the data from our primary region to secondary region(Outside Sweden) like `FRA1` within the same public cloud provider on a second object storage.

Reviewer 2: I would rephrase all the lines that are using Public and Compliant cloud to just `Infrastructure Provider` as this ADR is for all present and future Infrastructure Providers.
I would make it clear that for Infrastructure Providers that have a second region in Sweden or the Nordic Countries, we will rclone to that region, and where that is not available we will then rclone to another region that is outside Sweden within the same Infrastructure Provider.
If we need to be more specific on Infra provider names and regions then I think those should go in the `Recommendations for Platform Administrators` section.
@cristiklein what is your  take on this?

> 2. Use `rclone` to replicate the data from our primary region to secondary region(Outside Sweden) like `FRA1` within the same public Infrastructure Provider on a second object storage.

Reviewer: ```suggestion
2. Use `rclone` to replicate the data from our primary region to secondary region (outside Sweden) like `FRA1` within the same public Infrastructure Provider on a second object storage.
```

    - `Good`, because it supports our goals of data redundancy, resilience, disaster recovery preparedness, privacy, compliance, encrypted data transfer, and cross-region accessibility.

3. Use `rclone` to replicate the data from our `primary` Infrastructure Provider to `secondary` Compliant Infrastructure Provider on a second object storage.

    - We investigated this with Infrastructure Providers, but found that there is no good way to enable communication between public Infrastructure Provider object storage and compliant Infrastructure Provider object storage.

4. Do nothing and accept the risk of data loss.

    - `Bad`, because not implementing data replication leaves our services vulnerable to data loss in case of hardware failures, system errors, or accidental deletions, and recovery options become limited.

## Decision Outcome

Chosen option:

> - Use `rclone` to replicate the data from our primary region to secondary region(Outside Sweden) like `FRA1` within the same public Infrastructure Provider on a second object storage.

Reviewer: ```suggestion
- Use `rclone` to replicate the data from our primary region to secondary region (outside Sweden) like `FRA1` within the same public Infrastructure Provider on a second object storage.
```

> - Encrypt the backups before sending to an off-site location outside of Sweden and use the encryption feature in `rclone` which adheres to our [cryptography policy](../operator-manual/cryptography.md).

Reviewer 2: It feels to me that this section would fit better in the `Recommendations for Platform Administrators` section.
@cristiklein what is your take on this?

Reviewer: I agree @lucianvlad

Author: thanks @lucianvlad ,  done the changes.
